src/mono/mono/metadata/mono-debug.c

```diff
@@ -1066,81 +1067,60 @@ mono_is_debugger_attached (void)
-
 /**
  * mono_register_symfile_for_assembly:
  */
 void
 mono_register_symfile_for_assembly (const char *assembly_name, const mono_byte *raw_contents, int size)
 {
-	BundledSymfile *bsymfile;
-
-	bsymfile = g_new0 (BundledSymfile, 1);
-	bsymfile->aname = assembly_name;
-	bsymfile->raw_contents = raw_contents;
-	bsymfile->size = size;
-	bsymfile->next = bundled_symfiles;
-	bundled_symfiles = bsymfile;
-}
-
-static gboolean
-bsymfile_match (BundledSymfile *bsymfile, const char *assembly_name)
-{
-	if (!strcmp (bsymfile->aname, assembly_name))
-		return TRUE;
-#ifdef ENABLE_WEBCIL
-	const char *p = strstr (assembly_name, ".webcil");
-	/* if assembly_name ends with .webcil, check if aname matches, with a .dll extension instead */
-	if (p && *(p + strlen(".webcil")) == 0) {
-		size_t n = p - assembly_name;
-		if (!strncmp (bsymfile->aname, assembly_name, n)
-			&& !strcmp (bsymfile->aname + n, ".dll"))
-			return TRUE;
+	MonoBundledAssemblyResource *assembly_resource = mono_bundled_resources_get_assembly_resource (assembly_name);
+	if (!assembly_resource) {
+		assembly_resource = g_new0 (MonoBundledAssemblyResource, 1);
+		assembly_resource->resource.type = MONO_BUNDLED_ASSEMBLY;
+		assembly_resource->resource.id = assembly_name;
+		assembly_resource->resource.free_bundled_resource_func = &mono_bundled_resources_free_bundled_resource_func;
+	} else {
+		// Ensure the MonoBundledSymbolData has not been initialized
+		g_assert (!assembly_resource->symbol_data.data && assembly_resource->symbol_data.size == 0);
 	}
-#endif
-	return FALSE;
+	assembly_resource->symbol_data.data = (const uint8_t *)raw_contents;
+	assembly_resource->symbol_data.size = (uint32_t)size;
+	mono_bundled_resources_add ((MonoBundledResource **)&assembly_resource, 1);
 }
 
 static MonoDebugHandle *
 open_symfile_from_bundle (MonoImage *image)
 {
-	BundledSymfile *bsymfile;
-
-	for (bsymfile = bundled_symfiles; bsymfile; bsymfile = bsymfile->next) {
-		if (!bsymfile_match (bsymfile, image->module_name))
-			continue;
+	MonoBundledAssemblyResource *assembly = mono_bundled_resources_get_assembly_resource (image->module_name);
+	if (assembly && assembly->symbol_data.data)
+		return mono_debug_open_image (image, assembly->symbol_data.data, assembly->symbol_data.size);
 
-		return mono_debug_open_image (image, bsymfile->raw_contents, bsymfile->size);
+#ifdef ENABLE_WEBCIL
+	int len = strlen (image->module_name);
+	char *module_name_dll_suffix = strdup (image->module_name);
+	/* if image's module_name ends with .webcil, check if theres a bundled resource with a .dll extension instead */
+	if (module_name_dll_suffix && len >= 7 && !g_strcasecmp (".webcil", &image->module_name [len - 7])) {
+		memcpy (module_name_dll_suffix + len - 7, ".dll", 4);
+		*(module_name_dll_suffix + len - 3) = '\0';
+		assembly = mono_bundled_resources_get_assembly_resource (module_name_dll_suffix);
 	}
+	g_free (module_name_dll_suffix);
+	if (assembly && assembly->symbol_data)
```

Comment:
This cause compile errors on multiple wasm lanes:

```
  /__w/1/s/src/mono/mono/metadata/mono-debug.c:1108:15: error: invalid operands to binary expression ('MonoBundledAssemblyResource *' (aka 'struct _MonoBundledAssemblyResource *') and 'MonoBundledSymbolData' (aka 'struct _MonoBundledSymbolData'))
          if (assembly && assembly->symbol_data)
              ~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~
  1 error generated.

```

it should be:

`if (assembly && assembly->symbol_data.data)`